sklearn/datasets/_openml.py

```diff
@@ -37,7 +37,9 @@ def _get_local_path(openml_path: str, data_home: str) -> str:
     return os.path.join(data_home, "openml.org", openml_path + ".gz")
 
 
-def _retry_with_clean_cache(openml_path: str, data_home: Optional[str]) -> Callable:
+def _retry_with_clean_cache(
+    openml_path: str, data_home: Optional[str], parser: Optional[str] = None
```

Comment:
API wise, I think it's strange to pass the parser to `_retry_with_clean_cache`. What do you think of passing in a `no_retry_exception`:

```python
def _retry_with_clean_cache(, no_retry_exceptions=None):
    ...
    except Exception as exc:
        if no_retry_exceptions is not None and isinstance(exc, no_retry_exceptions):
            raise
```

And the caller:

```python
if parser == "pandas":
    from pandas.errors import ParserError
    no_retry_exceptions = ParserError
else:
    no_retry_exceptions = None

X, y, frame, categories = _retry_with_clean_cache(url, data_home, no_retry_exceptions)(
```